src/main.cpp

```diff
@@ -90,87 +91,201 @@ namespace
-                        return user_is_administrator(*rei.rsComm);
-                    }
-
-                    namespace adm = irods::experimental::administration;
-
-                    const adm::user user{rei.uoic->userName, rei.uoic->rodsZone};
-
-                    for (auto&& editor : config->at("editors")) {
-                        if (const auto& type = editor.at("type").get_ref<const std::string&>(); type == "group") {
-                            const adm::group group{editor.at("name").get_ref<const std::string&>()};
-
-                            if (adm::server::user_is_member_of_group(*rei.rsComm, group, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                        else if (type == "user") {
-                            if (editor.at("name").get_ref<const std::string&>() == adm::server::local_unique_name(*rei.rsComm, user)) {
-                                return CODE(RULE_ENGINE_CONTINUE);
-                            }
-                        }
-                    }
-
-                    // At this point, the user is not an administrator and they aren't a member of
-                    // the editors list. Therefore, we return an error because the user is attempting to
-                    // modify metadata in a guarded namespace.
-                    return ERROR(CAT_INSUFFICIENT_PRIVILEGE_LEVEL, "User is not allowed to modify metadata");
-                }
-            }
+		_exists = (_rule_name == "pep_api_mod_avu_metadata_pre") ||
+		          (_rule_name == "pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto list_rules(irods::default_re_ctx&, std::vector<std::string>& _rules) -> irods::error
+	{
+		_rules.push_back("pep_api_mod_avu_metadata_pre");
+		_rules.push_back("pep_api_atomic_apply_metadata_operations_pre");
+		return SUCCESS();
+	}
+
+	auto check_operation_for_violations(const std::string& _object_name, ruleExecInfo_t& rei) -> irods::error
+	{
+		const auto config = load_plugin_config(rei);
+        if(config == std::nullopt) {
+			return CODE(RULE_ENGINE_CONTINUE);
         }
-        catch (const json::exception&) {
-            // clang-format off
+		const auto prefixes_iter = config->find("prefixes");
+		if (prefixes_iter == config->cend()) {
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		const auto prefix_matched =
+			std::any_of(prefixes_iter->cbegin(), prefixes_iter->cend(), [&_object_name](const json& prefix) {
+				return boost::starts_with(_object_name, prefix.get_ref<const std::string&>());
+			});
+		if (!prefix_matched) {
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		const auto admin_iter = config->find("admin_only");
+		if (admin_iter == config->cend()) {
+			return CODE(RULE_ENGINE_CONTINUE);
+		}
+		if (admin_iter->get<bool>()) {
+			return user_is_administrator(*rei.rsComm);
+		}
```

Comment:
This is introducing a subtle but important change of behavior: `admin_only` would now be a required configuration with this. The existing behavior (if I'm reading it right) makes `admin_only` optional with a default of `false`. I think we want something like this if we want to maintain the existing behavior:
```suggestion
		if (admin_iter != config->cend() && admin_iter->get<bool>()) {
			return user_is_administrator(*rei.rsComm);
		}
```